src/sleep/sleep.c

```diff
@@ -263,41 +263,91 @@ static int execute(
+                        /* In case of no battery, system suspend interval will be set to hibernatedelaysec. */
+                else
+                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
+
+
+                log_debug("Set timerfd wake alarm for %s", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
+                /* Wake alarm for system with or without battery to hibernate or estimate discharge rate whichever is applicable */
+                timespec_store(&ts.it_value, suspend_interval);
+
+                r = timerfd_settime(tfd, 0, &ts, NULL);
+                if (r < 0)
+                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
 
-        timespec_store(&ts.it_value, sleep_config->hibernate_delay_sec);
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
 
-        r = timerfd_settime(tfd, 0, &ts, NULL);
-        if (r < 0)
-                return log_error_errno(errno, "Error setting hibernate timer: %m");
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+                /* Store fd_wait status */
+                woken_by_timer = FLAGS_SET(r, POLLIN);
+
+                r = read_battery_capacity_percentage();
+                if (r >= 0) {
+                        current_capacity = r;
+                        log_debug("Current battery charge percentage after wakeup: %d%%", current_capacity);
+                } else if (r == -ENOENT) {
+                        log_debug_errno(r, "Battery capacity percentage unavailable, cannot estimate discharge rate: %m");
+                        break;
+                        /* In case of no battery, system will be hibernated after 1st cycle of suspend */
+                } else
+                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
+
+                if (!woken_by_timer) {
+                        if (current_capacity < last_capacity) {
+                                /* We woke up before alarm time, estimate discharge rate using suspension duration. */
+                                after_timestamp = now(CLOCK_BOOTTIME);
+                                log_debug("Estimating discharge rate....");
+                                estimated_discharge_rate = ((last_capacity - current_capacity) * 60 * 60 * USEC_PER_SEC) / (after_timestamp - before_timestamp);
+                                /* The capacity difference is multiplied with 3600 to convert it to per hour */
+                                log_debug("Manual Wakeup. Battery discharge rate is %d%% per hour", estimated_discharge_rate);
+
+                        }
+                        return 0; /* return as manual wakeup done. This also will return in case battery was charged during suspension */
```

Comment:
Why do you calculate the discharge rate here? The value will never used... I guess this should be moved to the second commit. Or, squash the two commits.

Reply:
The commit are actually divided based on tasks mentioned in this : https://github.com/systemd/systemd/issues/21165

So to keep that. task 3 and 4 are in commit 1 and task 5 is in commit 2.